datasets/flwr_datasets/partitioner/image_semantic_partitioner.py

```diff
@@ -0,0 +1,505 @@
+
+    Parameters
+    ----------
+    num_partitions : int
+        The total number of partitions that the data will be divided into.
+    partition_by : str
+        Column name of the labels (targets) based on which Dirichlet sampling works.
+    efficient_net_type: int
+        The type of pretrained EfficientNet model.
+        Options: [0, 1, 2, 3, 4, 5, 6, 7], corresponding to EfficientNet B0-B7 models.
+    batch_size: int
+        The batch size for EfficientNet extracting embeddings.
+    pca_components: int
+        The number of PCA components for dimensionality reduction.
+    gmm_max_iter: int
+        The maximum number of iterations for the GMM algorithm.
+    gmm_init_params: str
+        The initialization method for the GMM algorithm.
+        Options: ["random", "kmeans", "k-means++"]
+    use_cuda: bool
+        Whether to use CUDA for computation acceleration.
+    shuffle: bool
+        Whether to randomize the order of samples. Shuffling applied after the
+        samples assignment to partitions.
+    sample_seed: int
+        Seed used for sampling.
+    pca_seed: Optional[int]
+        Seed used for PCA dimensionality reduction. If not set, sample_seed is used.
+    gmm_seed: Optional[int]
+        Seed used for GMM clustering. If not set, sample_seed is used.
+    image_column_name: Optional[str]
+        The name of the image column in the dataset. If not set, the first image column
+        is used.
+
+
+    Examples
+    --------
+    >>> from flwr_datasets import FederatedDataset
+    >>> from flwr_datasets.partitioner import ImageSemanticPartitioner
+    >>> partitioner = ImageSemanticPartitioner(
+    >>>     num_partitions=10,
+    >>>     partition_by="label",
+    >>>     pca_components=128,
+    >>>     gmm_max_iter=100,
+    >>>     gmm_init_params="kmeans",
+    >>>     use_cuda=True,
+    >>>     shuffle=True,
+    >>> )
+    >>> fds = FederatedDataset(dataset="mnist", partitioners={"train": partitioner})
+    >>> partition = fds.load_partition(0)
+    >>> print(partition[0])  # Print the first example
+    {'image': <PIL.PngImagePlugin.PngImageFile image mode=L
+    size=28x28 at 0x7FCF49741B10>, 'label': 3}
+    >>> partition_sizes = partition_sizes = [
+    >>>     len(fds.load_partition(partition_id)) for partition_id in range(5)
+    >>> ]
+    >>> print(sorted(partition_sizes))
+    [3163, 5278, 5496, 6320, 9522]
+    """
+
+    def __init__(  # pylint: disable=R0913
+        self,
+        num_partitions: int,
+        partition_by: str,
+        efficient_net_type: int = 3,
+        batch_size: int = 32,
+        pca_components: int = 128,
+        gmm_max_iter: int = 100,
+        gmm_init_params: str = "kmeans",
+        use_cuda: bool = False,
+        shuffle: bool = True,
+        sample_seed: int = 42,
+        pca_seed: Optional[int] = None,
+        gmm_seed: Optional[int] = None,
+        image_column_name: Optional[str] = None,
```

Comment:
According to, unfortunately, unwritten convention, let's keep the shuffling parameters last and move this parameter anywhere before the shuffle.